OpenSSL-Strategy.md

```diff
@@ -0,0 +1,361 @@
+  Fedora and the majority of other Linux distributions follow a standard policy
+  of separating dependencies and dynamically linking wherever possible.  The
+  Node.js packages on these systems are linked via the packaging dependency
+  mechanisms to OpenSSL 1.0.2 packages and the `node` binaries that ship by
+  default on these platforms use the shared OpenSSL 1.0.2 library installed by
+  those packages.  Therefore, the floating patches do not apply and it is
+  possible that a different version of OpenSSL 1.0.2 is in use by Node.js than
+  the version that was shipped in the source tree.
+* **NodeSource Builds**: Binaries
+  [distributed](https://github.com/nodesource/distributions) by NodeSource are
+  heavily used, and do not follow the Linux distribution policyof dependency
+  seperation. NodeSource uses the Node.js defaults for `configure` and compile
+  and statically link in the OpenSSL source shipped bundled with Node.js.
+* **Homebrew**: The
+  [formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/node.rb)
+  for Node.js as distributed on macOS by the popular `brew` command, by default,
+  will compile using the default `configure` configuration. It is possible to
+  override this using a `with-openssl` option which will compile against the
+  version of OpenSSL that was most recently installed with `brew` but this is
+  not believed to be in common use.
+
+
+## Node.js version-specific strategy
+
+### Node.js version 4.x (EOL 2018-04-30)
+
+No longer maintained.  Not discussed further.
+
+### Node.js versions 6.x (EOL 2019-04-30) and 8.x (EOL 2019-12-31)
+
+Node.js 6.x and 8.x have the same OpenSSL versions and policies.
+
+Node.js 8.x is unusual because it's end-of-life would normally be April, 2020,
+but it was moved earlier to conincide with the end-of-life of OpenSSL 1.0.2.
+
+* OpenSSL version: 1.0.2 (kept up-to-date with latest)
+* Allowed shared OpenSSL version: 1.0.2x
+* Default minimum TLS version is TLSv1, default maximum is TLSv1.2. No other
+  TLS versions are supported by this version of OpenSSL or Node.js.
+* FIPS: supports being configured to build against OpenSSL FIPS 2.0. FIPS
+  configuration is not default, and OpenSSL FIPS 2.0 is not included in
+  Node.js.
+
+XXX Do we support FIPS builds, or do we treat like other shared builds, best
+effort to fix but not release critical if its broken?
+
+
+### Node.js version 10.x (EOL April-2021)
+
+OpenSSL LTS support timing, the lack of OpenSSL LTS planning and the lack of a
+clear timeframe for a new FIPS module complicates Node.js 10.
+
+For Node.js < 10.16.0:
+* OpenSSL version: 1.1.0
+* Allowed shared OpenSSL version: 1.0.2 or 1.1.0
+* Default minimum TLS version is TLSv1, default maximum is TLSv1.2. No other
+  TLS versions are supported by this version of OpenSSL or Node.js.
+* FIPS: not supported
+
+For Node.js >= 10.16.0:
+* OpenSSL version: 1.1.1
+* Allowed shared OpenSSL version: 1.0.2 or 1.1.0 or 1.1.1
```

Comment:
I have to confirm that 1.0.2 is possible. I assume that any shared builds that worked with 10.15.3 have to work with 10.16.0?

Reply:
I don't think we should break 1.0.2 support in 10.x.  If its a matter or not being able to get 1.1.1 in without doing that then we need to discuss.

Reply:
Confirmed 1.0.2 isn't supported by v10.x. 1.1.0 is. I updated text.

Reply:
If 10.x supports 1.0.2 then we should continue to support it. I built v10.x in ci, https://ci.nodejs.org/job/node-test-commit-linux-containered/11905/, and it built and passed against 1.0.2r, but my local attempts to build against shared openssl 1.0.2q fail completely, way too many missing symbols. I'm retrying with openssl 1.0.2r, though since patch releases shouldn't add new APIs, I'm mystified.

Reply:
> 
> 
> If 10.x supports 1.0.2 then we should continue to support it. I built v10.x in ci, https://ci.nodejs.org/job/node-test-commit-linux-containered/11905/, and it built and passed against 1.0.2r, but my local attempts to build against shared openssl 1.0.2q fail completely, way too many missing symbols. I'm retrying with openssl 1.0.2r, though since patch releases shouldn't add new APIs, I'm mystified.

er the build you've linked to skips over openssl 1.0.2 (and 1.1.1 and fips):
![image](https://user-images.githubusercontent.com/5445507/55417697-1b551a80-5569-11e9-99f0-0294f4342623.png)

Reply:
No, I can't build against 1.0.2r either. I'm retrying with a different openssl build configuration. @rvagg any comment? I even grepped around, AFAICT node_crypo.cc uses symbols not defined by openssl 1.0.2. I might need access to the docker containers to see what's going on.